File: FindXthNodeFromEnd/FindXthNodeFromEnd.cpp
// FindXthNodeFromEnd.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    if (pHead == nullptr)
        return;

    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

ListNode* GetNodeByIndex(ListNode* pHead, int index)
{
    if (index == -1)
        return nullptr;

    for (int i = 0; i < index; ++i)
    {
        pHead = pHead->next;
    }

    return pHead;
}

// Make the function recursive, so that it only traverse the list once
//  not sure if this is the fastest. Another solution would be count the
//  the total element in the list, then find the right node. Will need to
//  benchmark the two to figure out which solution is better. Maybe the
//  second is better, since it won't need waste time on prepairing the
//  parameters when calling this function
ListNode* XthNodeFromEnd(ListNode* head, int& x /*to use recursive, this value needs to be changed*/)
{
    if (head == nullptr)
    {   // start conting backword, need to clean the return value pNode
        return nullptr;
    }

    ListNode* pTemp = XthNodeFromEnd(head->next, x);
    if (pTemp != nullptr)   // if the node has been found, return it
        return pTemp;
    if (--x == 0)           // the node has not been found, need to count down
    {                       // here we go! we got it. need to store the node into
                            //  the return value
        return head;
    }

    // the node has not been found yet, return nullptr
    return nullptr;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count <= 0)
            break;

        ListNode* pHead = nullptr;
        ListNode* pTail = GetListFromInput(&pHead, count);

        std::cout << "Enter number n to remove the nth element from the end: ";
        int index = -1;
        std::cin >> index;

        int xth = index;
        ListNode* pNode = XthNodeFromEnd(pHead, index);

        std::cout << "The " << xth << "th node from the end is " << (pNode ? pNode->data : -1);
        std::cout << std::endl;

        FreeList(pHead);
    }
}

File: Go/FindXthNodeFromEnd.go
package main

import (
	"fmt"
)

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) (*ListNode, *ListNode) {
	if len(v) == 0 {
		return nil, nil
	}

	if len(v) == 1 {
		pNode := &ListNode{v[0], nil}
		return pNode, pNode
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead, pCur
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
	}
	if pNode.next != nil {
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func FindNode(head *ListNode, x int) (*ListNode, int) {
	if head == nil {
		return head, 0
	}

	pNode, i := FindNode(head.next, x)
	if pNode != nil {
		return pNode, i
	}

	i++
	if i == x {
		return head, i
	}

	return nil, i
}

func XthNodeFromEnd(head *ListNode, x int) *ListNode {
	pNode, _ := FindNode(head, x)

	return pNode
}

func main() {
	for {
		fmt.Print("Number of nodes: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the list: ")
		pNode, _ := GetListFromArray(vec)

		fmt.Print("Please enter number x: ")
		var x int
		fmt.Scan(&x)

		p := XthNodeFromEnd(pNode, x)
		if p != nil {
			fmt.Printf("The %dth node form end of the linked list is ", x)
			fmt.Println(XthNodeFromEnd(pNode, x).data)
		} else {
			fmt.Printf("There are less than %d nodes in the list", x)
		}
	}
}
